submissions/nadsatt/post-oop/index.js

```diff
@@ -0,0 +1,152 @@
+   - FOR I: as i see, there is no unused/unrelevant superclass' methods/props in subclasses
+   - FOR D: -
+*/ 
+
+class Inhabitant {
+   constructor(species, saying, gender, name, friends = []){
+      this.species = species;
+      this.saying = saying;
+      this.gender = gender;
+      this.name = name;
+      this.friends = friends;
+   }
+
+   toString(){
+      return `species: ${this.species}, saying: ${this.saying}, gender: ${this.gender}, name: ${this.name}, friends: ${new FriendsManager(this).getFriends()}`;
+   }
+}
+
+class FriendsManager {
+   constructor(inhabitant){
+      this.inhabitant = inhabitant;
+   }
+
+   addFriends(...friends){
+      this.inhabitant.friends.push(...friends);
+   }
+
+   getFriends(){
+      return this.inhabitant.friends.map(friend => friend.name).join(' ') || 'no friends';
+   }
+}
+
+class Primate extends Inhabitant {
+   constructor(species, saying, gender, name, ...friends){
+      super(species, saying, gender, name, friends);
+      this.hands = 2;
+      this.legs = 2;
+   }
+
+   toString(){
+      return super.toString() + `, legs: ${this.legs}, hands: ${this.hands}`;
+   }
+}
+
+class Human extends Primate {
+   constructor(saying, gender, name, ...friends){
+      super('human', saying, gender, name, ...friends);
+   }
+}
+
+class Man extends Human {
+   constructor(name, ...friends){
+      super('i\'m a man!', 'male', name, ...friends);
+   }
+}
+
+class Woman extends Human {
+   constructor(name, ...friends){
+      super('i\'m a woman!', 'female', name, ...friends);
+   }
+}
+
+class Predator extends Inhabitant {
+   constructor(species, saying, gender, name, ...friends){
+      super(species, saying, gender, name, friends);
+      this.legs = 4;
+   }
+
+   toString(){
+      return super.toString() + `, legs: ${this.legs}`;
+   }
+}
+
+class Dog extends Predator {
+   constructor(gender, name, ...friends){
+      super('dog', 'woof-woof!', gender, name, ...friends);
+   }
+}
+
+class Cat extends Predator {
+   constructor(gender, name, ...friends){
+      super('cat', 'meov-meov!', gender, name, ...friends);
+   }
+}
+
+class CatWoman extends Cat {
+   constructor(name, ...friends){
+      super('female', name, ...friends);
+      this.species = 'human';
+      this.hands = 2;
+      this.legs = 2;
+   }
+
+   toString(){
+      return super.toString() + `, hands: ${this.hands}`;
+   }
+}
+
+const cat1 = new Cat('female', 'Kitty');
+const dog1 = new Dog('male', 'Bob', cat1);
+const man1 = new Man('Nick');
+const woman1 = new Woman('Lucy');
+const woman2 = new Human('i\'m a special woman!', 'female', 'Linda');
+const catWoman1 = new CatWoman('Kate', man1, woman2);
+
+const cat1Friends = new FriendsManager(cat1);
+cat1Friends.addFriends(dog1);
+
+const man1Friends = new FriendsManager(man1);
+man1Friends.addFriends(catWoman1);
+
+const woman2Friends = new FriendsManager(woman2);
+woman2Friends.addFriends(catWoman1);
```

Comment:
**FIX REQUIRED**

If you had FriendsManager instantiated in Inhabitant constructor you would not need to create a new FriendsManager object every time you need to add friends.
Save some memory, take care of JS garbage collector.